dxr/plugins/python/indexers.py

```diff
@@ -249,44 +257,43 @@ def analyze_tokens(self):
-                    continue
-
-                if tok_name in ('def', 'class'):
-                    node_type, node_start = 'definition', (utf8_start, start)
-                    continue
-
-                # Record all name nodes in the token table.  Currently unused,
-                # but will be needed for recording variable references.
-                node_start_table[utf8_start] = (start, end)
-                node_type, node_start = 'name', (utf8_start, start)
+                # names, we start the queue for the current parenthesis level
+                # at the byte offset for the keyword token, but only start
+                # pushing character offsets once we're past the keyword.
+                paren_stack.setdefault(paren_level, (utf8_start, []))
+
+                if tok_name not in ('def', 'class'):
+                    paren_stack[paren_level][1].append((start, end))
+
+                continue
 
             elif tok_type == token.OP:
-                # In order to properly capture the start and end of function
-                # calls, we need to keep track of the parens.  Put the
-                # starting positions on a stack (here implemented with a dict
-                # so that it can be sparse), but only if the previous node was
-                # a name.
-                if tok_name == '(':
-                    if node_type == 'name':
-                        paren_stack[paren_level] = node_start
+                # Container delimiters (parens, brackets, and braces) start a
+                # new context where the node for following name tokens will no
+                # longer be tied to the position of the head of the current
+                # queue.  So, keep track of the current context with a stack,
+                # here implemented with a dict so that it can be sparse.
+                if tok_name in '([{':
                     paren_level += 1
-                elif tok_name == ')':
+                elif tok_name in '}])':
+                    node_start, node_queue = paren_stack.pop(paren_level, (None, None))
+                    if node_start is not None:
+                        node_start_table[node_start] = node_queue
                     paren_level -= 1
-                    if paren_level in paren_stack:
-                        call_start = paren_stack.pop(paren_level)
-                        call_start_table[call_start[0]] = (call_start[1], end)
+                elif tok_name == '.':
+                    # Attribute access.  Don't reset, stay at the same level.
+                    pass
+                else:
+                    node_start, node_queue = paren_stack.pop(paren_level, (None, None))
+                    if node_start is not None:
+                        node_start_table[node_start] = node_queue
 
-                node_type, node_start = None, None
+            elif tok_type == token.NEWLINE:
+                node_start, node_queue = paren_stack.pop(paren_level, (None, None))
```

Comment:
This is the 3rd time this is repeated. I wonder if there's a clever way to factor it out.